### ISSUE TITLE
Profie fields not checking permissions correctly

### ISSUE THREAD

**Issue by Destroy666x** [mybb/mybb#1373]
Yet another feature that is not checked correctly for permissions - if a user has 5 posts and fills a profile field, which will be later set by admin to minimum 10 posts, it will still display in postbit/profile for that user.

EDIT: same applies to `editableby` in both profile/postbit. And the Display on profile doesn't seem to accept 0.

EDIT2: postbit doesn't check `viewableby` either.. And empty fields need to be hidden there.

**Comment by Sama34** [comment]
Please note:
http://community.mybb.com/thread-168610-post-1148133.html#pid1148133

I think this need to be fixed in 1.8.5 so if you can;'t work on this feel free to assign me.

**Comment by Destroy666x** [comment]
I'll submit several PRs including this tommorow/on Sunday.

**Comment by Sama34** [comment]
@Destroy666x

**Comment by Destroy666x** [comment]
Sorry, forgot about this, preparing a PR.

But not quite sure what do you mean by this post and where exactly it applies: http://community.mybb.com/thread-168610-post-1148140.html#pid1148140 What are the reproduction steps?

**Comment by Sama34** [comment]
Just add a regex to the bio field and make it not required (say, regex to require only digits). Then go to the ACP and leave the field empty. You wouldn't be able to update the user.

```PHP
if(!empty($profilefield['regex']) && !preg_match("#".$profilefield['regex']."#i", $profile_fields[$field]))
{
	$this->set_error('bad_profile_field_value', array($profilefield['name']));
}
```

`$profilefield['editableby'] = -1;` shouldn't be set like this. If `$this->data['profile_fields_editable']` is true the field should be saved regardless of the value.

**Comment by Sama34** [comment]
Also note that if the field is not editable by anyone it will not be shown yet the error is displayed.

**Comment by Sama34** [comment]
>  same applies to `editableby`

No, it doesn't. You are assuming that an administrator is not going to create a profile field that is only editable by moderators, which should be displayed regardless of permissions.

You will break functionality changing this behavior.

> if a user has 5 posts and fills a profile field, which will be later set by admin to minimum 10 posts, it will still display in postbit/profile for that user.

Pretty much similar to above.

**Comment by Destroy666x** [comment]
And where does it say that profile fields have to be always displayed for moderators/administrators on profile/postbit regardless of other permissions? You're missing these points:
1. Without checking `editableby` (and `numposts`) at all admin can't change limitations properly later, nor move users between groups, which is an obvious bug - for example user can fill a field while being in group X, then get moved to group Y which shouldn't be able to edit the field and without the code I added the field would be still displayed (and not only to moderators/admins).
2. The previous profile code which displayed all fields to admins/mods was totally wrong because even `viewableby` may skip some groups that are moderators.
3. Other stuff, such as signatures and websites, is handled the same way:
https://github.com/mybb/mybb/blob/feature/inc/functions_post.php#L346
https://github.com/mybb/mybb/blob/feature/inc/functions_post.php#L724
No admin/mod checks either.

What do others think? @Stefan-ST @JN-Jones @ATofighi @euantorano (can't tag @mybb/developers anymore...).

**Comment by Sama34** [comment]
> where does it say that profile fields editable by mods/admins have to be always displayed on profile/postbit regardless of other permissions?

> may skip some groups that are moderators

The `viewableby` implies this (as long as the mod group is selected of course). The `editableby` only handles _editing_ permissions [1].

> Other stuff with "who can edit" permission, such as signatures and websites, is handled the same way:

This is the expected behavior regardless of _similar_ fields behaving slightly differently. TBH I don't see why it was accepted to modify the behavior of the website and signature fields either [2].

Since, again, viewable and editable permissions are not the same [1].

> (and `numposts`) at all admin can't change limitations properly later, nor move users between groups, which is an obvious bug

Analise the following possibility:
1. Administrator creates a group promotion that requires `10` amount of posts.
2. User gets promoted.
3. Administrator updates the group promotion to require now `11` posts instead.
4. Users promoted before the update keep their promotion.

By your logic this will be considered a _bug_. It is not.

> And I doubt there is any other solution

What you are suggesting here should rather be handled by new settings (like a secondary viable permission setting or whatever).

So:
- [1] I as the administrator should be able to create a field that is only editable during registration by users but not after it.
- [2] I as an administrator should be able to edit and view a field regardless of the user being able to edit that field.

**Comment by Destroy666x** [comment]
> The `editableby` only handles editing permissions [1].

So you're basically stating that at the moment there is no "Group XYZ can use this field" permission which would be a lot more useful for admins than `editableby` (with your definition of it at least) and it should be added (see below)?

> Analise the following possibility ... By your logic this will be considered a bug. It is not.

No, by my logic it's a wrong comparison. Promotions are supposed to be a single-time process and any admin can always undo them with another "reverse promotion" (or in other words - demotion) if that's needed.

With your example:
1. Add New Promotion applying only to the XYZ group, change primary group to any other.
2. Set Post Count to Less Than 11.
3. Run it. Done, no problems at all. It only wouldn't work for Time Registered and Time Online, which are the only fields that don't have a Greater/Less option for some reason, thus it's another weird implementation, in my opinion.

Fields on the other hand are viewable/editable by some groups constantly and as I said there is no way to change limitations for them except by recreating the field or running queries in phpMyAdmin.

> I as an administrator should be able to edit and view a field regardless of the user being able to edit that field.

And you can always do that in ACP. But front-end should show the actual state of the field - if a group can't use it, admin should see that the user being only in it can't use it.

> What you are suggesting here should rather be handled by new settings

Here I agree, it could actually be partially solved by adding a `usableby` or whatever groupselect field. Nothing against adding it and checking it instead, but I'll wait for other opinions on this.

**Comment by Sama34** [comment]
@mybb/developers

**Comment by JN-Jones** [comment]
As I haven't worked with profilefields lately I may be wrong, but from what I've read the discussion is about when and where somebody is allowed to edit a profilefield? IMHO it should be like it's done with other fields:
- UCP: only if user has permission to do so, even if the user is admin
- MCP/ACP: regardless of permission.

And the permissions should always use the current settings, if the admin changes them they should be updated for existing profilefields too.

The case @Sama34 pointed out should be handled by a new setting (which should probably be added in a later version instead of 1.8.5).

**Comment by Sama34** [comment]
> when and where somebody is allowed to edit a profilefield?

It is about whether the `editableby` field should only be used for editing (how it currently is) and the `viewableby` field for viewable permissions. I implemented those new fields on 1.8 envisioning the functionality how it currently is (assuming not bugs are present, which there are, at least with the `viewableby` field which is pretty much the point of this issue).

UserCP and ModCP work as you mentioned. It is about display on profile, postbit, etc.

There is also an issue with visibility of the field while editing a field in the UCP, MCP, and IIRC even ACP.

http://community.mybb.com/thread-168610-post-1148140.html#pid1148140

> And the permissions should always use the current settings, if the admin changes them they should be updated for existing profilefields too.

I'm fine with that for the post count and `viewableby` permission field.

**Comment by Ben-MyBB** [comment]
I will give this a shot, it should be relatively straight forward after adding the permissions check in before the fields get displayed on the post bit & profile.

**Comment by Ben-MyBB** [comment]
@arcalys maybe?

**Comment by arcalys** [comment]
> [@arcalys](https://github.com/arcalys) maybe?

I can have a look, but I’m not sure I fully understand the scenarios to fix. Could you clarify?

**Comment by Sama34** [comment]
I think there are various custom profile fields settings (group permissions, post requirement, etc) which may be checked during edit or display, but not both.

This causes unexpected behaviour and in some cases make the feature dysfunctional (my linked post above), like displaying fields value even if the user does not meet the post requirement, while at the same time blocking the user from updating such fields, etc.

Regards.
